### PR TITLE
Proto for the NFT issuance

### PR DESCRIPTION
**Pull request by dzmitryhil**
This change is [Reviewable](https://reviewable.io/reviews/coreumfoundation/coreum/274)

### CODE REVIEW

**Review by ysv — CHANGES_REQUESTED** [review]
> Reviewed 3 of 3 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/coreumfoundation/coreum/274)* status: all files reviewed, 1 unresolved discussion (waiting on @dzmitryhil, @miladz68, and @wojtek-coreum)

___
*[`proto/coreum/asset/v1/tx.proto` line 73 at r1](https://reviewable.io/reviews/coreumfoundation/coreum/274#-NHyGa70ESl3hOWJ2jF3:-NHyGa70ESl3hOWJ2jF4:bpycrsy) ([raw file](https://github.com/coreumfoundation/coreum/blob/b4174e838dbe4cdd99d9414ea359f6634b80f738/proto/coreum/asset/v1/tx.proto#L73)):*
> ```protobuf
> 
> // NonFungibleToken is a full representation of the non fungible token.
> message MsgIssueNonFungibleToken {
> ```

lets discuss your questions on the call
Another notes I have:
1. can user create class ?
2. do we need to include class in NFT issuance
3. it is really important to have bulk NFT issuance because creating them one by one is inconvenient since only 1 tx could be included in a block

**Review by dzmitryhil — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/coreumfoundation/coreum/274)* status: all files reviewed, 1 unresolved discussion (waiting on @miladz68, @wojtek-coreum, and @ysv)

___
*[`proto/coreum/asset/v1/tx.proto` line 73 at r1](https://reviewable.io/reviews/coreumfoundation/coreum/274#-NHyGa70ESl3hOWJ2jF3:-NHyHaetE_u3N4oAyglL:b-irww89) ([raw file](https://github.com/coreumfoundation/coreum/blob/b4174e838dbe4cdd99d9414ea359f6634b80f738/proto/coreum/asset/v1/tx.proto#L73)):*
<details><summary><i>Previously, ysv (Yaroslav Savchuk) wrote…</i></summary><blockquote>

lets discuss your questions on the call
Another notes I have:
1. can user create class ?
2. do we need to include class in NFT issuance
3. it is really important to have bulk NFT issuance because creating them one by one is inconvenient since only 1 tx could be included in a block
</blockquote></details>

1. NFT issuance  = Class Creation
2. Bulk NFT issuance = Bulk Minting

**Review by dzmitryhil — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/coreumfoundation/coreum/274)* status: all files reviewed, 1 unresolved discussion (waiting on @miladz68, @wojtek-coreum, and @ysv)

___
*[`proto/coreum/asset/v1/tx.proto` line 73 at r1](https://reviewable.io/reviews/coreumfoundation/coreum/274#-NHyGa70ESl3hOWJ2jF3:-NHyHoYO9XkRxyjhhWQF:bs08jhw) ([raw file](https://github.com/coreumfoundation/coreum/blob/b4174e838dbe4cdd99d9414ea359f6634b80f738/proto/coreum/asset/v1/tx.proto#L73)):*
<details><summary><i>Previously, dzmitryhil (Dzmitry Hil) wrote…</i></summary><blockquote>

1. NFT issuance  = Class Creation
2. Bulk NFT issuance = Bulk Minting 
</blockquote></details>

Means:
1) We issue the NFT = Create a class (equal to the ERC-721 contract creation)
2) We mint the NFT in the class

File: proto/coreum/asset/v1/asset.proto [hunk 1]
```diff
@@ -30,3 +30,18 @@ message FungibleToken {
   string description = 4;
   repeated FungibleTokenFeature features = 5;
 }
+
+// NonFungibleToken is a full representation of the non fungible token.
+message NonFungibleToken {
```
Comment (dzmitryhil) [inline proto/coreum/asset/v1/asset.proto:35]: NonFungibleTokenClass

**Review by miladz68 — CHANGES_REQUESTED** [review]
> Reviewed 2 of 3 files at r1, 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/coreumfoundation/coreum/274)* status: all files reviewed, 8 unresolved discussions (waiting on @dzmitryhil, @wojtek-coreum, and @ysv)

___
*[`proto/coreum/asset/v1/asset.proto` line 35 at r2](https://reviewable.io/reviews/coreumfoundation/coreum/274#-NI0upnL7PWWoGJM2OdS:-NI0upnL7PWWoGJM2OdT:bg2b4tz) ([raw file](https://github.com/coreumfoundation/coreum/blob/53effc4d0cbf7646cdf1a76b0ba7530469d16b4a/proto/coreum/asset/v1/asset.proto#L35)):*
> ```protobuf
> 
> // NonFungibleToken is a full representation of the non fungible token.
> message NonFungibleToken {
> ```

This class definition already exists in the original nft module, it only lacks the issuer. so may be instead of adding the issuer to the structure, we can add a query to get the issuer ?
___
*[`proto/coreum/asset/v1/query.proto` line 29 at r2](https://reviewable.io/reviews/coreumfoundation/coreum/274#-NI0xMp_EyNZwF74qp1r:-NI0xMp_EyNZwF74qp1s:b-ou8b32) ([raw file](https://github.com/coreumfoundation/coreum/blob/53effc4d0cbf7646cdf1a76b0ba7530469d16b4a/proto/coreum/asset/v1/query.proto#L29)):*
> ```protobuf
> 
>   // NonFungibleToken queries the noe fungible token of the module.
>   rpc NonFungibleToken(QueryNonFungibleTokenRequest) returns (QueryNonFungibleTokenResponse) {
> ```

non-blocking
I left a comment above to remove NonFungibleToken type. if agreed,  then it follows that this query should be replaced with a query to get the issuer of the class and a query to list classess by issuer
___
*[`proto/coreum/asset/v1/tx.proto` line 100 at r2](https://reviewable.io/reviews/coreumfoundation/coreum/274#-NI0zjv61Gv_lXQI4yPB:-NI0zjv61Gv_lXQI4yPC:b-tgwbzf) ([raw file](https://github.com/coreumfoundation/coreum/blob/53effc4d0cbf7646cdf1a76b0ba7530469d16b4a/proto/coreum/asset/v1/tx.proto#L100)):*
> ```protobuf
>   string class_id = 1;
>   // id is a unique identifier of the NFT
>   // string id = 2;
> ```

non-blocking
I still think that letting the users provide IDs makes developers lives much simpler.
An example is that I want to mint 3 NFTs from inside smart contract and later update their data field according to some logic. In this I need to listen for events from smart contracts to know the ID and then store them, right ?

**Review by dzmitryhil — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/coreumfoundation/coreum/274)* status: all files reviewed, 8 unresolved discussions (waiting on @miladz68, @wojtek-coreum, and @ysv)

___
*[`proto/coreum/asset/v1/asset.proto` line 35 at r2](https://reviewable.io/reviews/coreumfoundation/coreum/274#-NI0upnL7PWWoGJM2OdS:-NI1JHcr64I0iQEVZEDn:b-8i7ik9) ([raw file](https://github.com/coreumfoundation/coreum/blob/53effc4d0cbf7646cdf1a76b0ba7530469d16b4a/proto/coreum/asset/v1/asset.proto#L35)):*
<details><summary><i>Previously, miladz68 (milad) wrote…</i></summary><blockquote>

This class definition already exists in the original nft module, it only lacks the issuer. so may be instead of adding the issuer to the structure, we can add a query to get the issuer ?
</blockquote></details>

Right we will reuse the queries from it
___
*[`proto/coreum/asset/v1/query.proto` line 29 at r2](https://reviewable.io/reviews/coreumfoundation/coreum/274#-NI0xMp_EyNZwF74qp1r:-NI1JNG146xpZZTg85wS:b-ginqso) ([raw file](https://github.com/coreumfoundation/coreum/blob/53effc4d0cbf7646cdf1a76b0ba7530469d16b4a/proto/coreum/asset/v1/query.proto#L29)):*
<details><summary><i>Previously, miladz68 (milad) wrote…</i></summary><blockquote>

non-blocking
I left a comment above to remove NonFungibleToken type. if agreed,  then it follows that this query should be replaced with a query to get the issuer of the class and a query to list classess by issuer
</blockquote></details>

Yes, we can remove them fully.
___
*[`proto/coreum/asset/v1/tx.proto` line 100 at r2](https://reviewable.io/reviews/coreumfoundation/coreum/274#-NI0zjv61Gv_lXQI4yPB:-NI1JQaT8cRLyAlAu3n1:b-h10arg) ([raw file](https://github.com/coreumfoundation/coreum/blob/53effc4d0cbf7646cdf1a76b0ba7530469d16b4a/proto/coreum/asset/v1/tx.proto#L100)):*
<details><summary><i>Previously, miladz68 (milad) wrote…</i></summary><blockquote>

non-blocking
I still think that letting the users provide IDs makes developers lives much simpler.
An example is that I want to mint 3 NFTs from inside smart contract and later update their data field according to some logic. In this I need to listen for events from smart contracts to know the ID and then store them, right ? 
</blockquote></details>

I see the pross of providing ID by users, but there are move cons IMHO.

**Review by miladz68 — DISMISSED** [review]
> Reviewed 4 of 4 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/coreumfoundation/coreum/274)* status: all files reviewed, 3 unresolved discussions (waiting on @dzmitryhil, @wojtek-coreum, and @ysv)

___
*[`proto/coreum/asset/v1/tx.proto` line 100 at r2](https://reviewable.io/reviews/coreumfoundation/coreum/274#-NI0zjv61Gv_lXQI4yPB:-NI1Rep9CJxvE2gNY4ok:b7qfayq) ([raw file](https://github.com/coreumfoundation/coreum/blob/53effc4d0cbf7646cdf1a76b0ba7530469d16b4a/proto/coreum/asset/v1/tx.proto#L100)):*
<details><summary><i>Previously, dzmitryhil (Dzmitry Hil) wrote…</i></summary><blockquote>

I see the pross of providing ID by users, but there are move cons IMHO.
</blockquote></details>

I will mark this as non-blocking

**Review by miladz68 — CHANGES_REQUESTED** [review]
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/coreumfoundation/coreum/274)* status: 3 of 4 files reviewed, 4 unresolved discussions (waiting on @dzmitryhil, @wojtek-coreum, and @ysv)

___
*[`proto/coreum/asset/v1/tx.proto` line 93 at r4](https://reviewable.io/reviews/coreumfoundation/coreum/274#-NI2B0ukEElgpqfwbf8u:-NI2B0ul5_ToEv2Fg-c4:b-31xy57) ([raw file](https://github.com/coreumfoundation/coreum/blob/228dde7a4a98df4cb93d5f34f7a24e853b65be06/proto/coreum/asset/v1/tx.proto#L93)):*
> ```protobuf
> // MsgMintNonFungibleToken defines message for the MintNonFungibleToken method.
> message MsgMintNonFungibleToken {
>   option (gogoproto.equal) = false;
> ```

I don't think we need this option

**Review by dzmitryhil — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/coreumfoundation/coreum/274)* status: 3 of 4 files reviewed, 3 unresolved discussions (waiting on @miladz68, @wojtek-coreum, and @ysv)

___
*[`proto/coreum/asset/v1/tx.proto` line 93 at r4](https://reviewable.io/reviews/coreumfoundation/coreum/274#-NI2B0ukEElgpqfwbf8u:-NI2GYVb904BHQcr_xVo:b-896fix) ([raw file](https://github.com/coreumfoundation/coreum/blob/228dde7a4a98df4cb93d5f34f7a24e853b65be06/proto/coreum/asset/v1/tx.proto#L93)):*
<details><summary><i>Previously, miladz68 (milad) wrote…</i></summary><blockquote>

I don't think we need this option
</blockquote></details>

Done.

**Review by miladz68 — CHANGES_REQUESTED** [review]
> Reviewed 2 of 2 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/coreumfoundation/coreum/274)* status: all files reviewed, 3 unresolved discussions (waiting on @dzmitryhil, @wojtek-coreum, and @ysv)

___
*[`proto/coreum/asset/v1/asset.proto` line 40 at r5](https://reviewable.io/reviews/coreumfoundation/coreum/274#-NI6QsLWAqLxLILeQkun:-NI6QsLWAqLxLILeQkuo:b-78i614) ([raw file](https://github.com/coreumfoundation/coreum/blob/fd2df96830dd7929d7e913a9bb30d2abf5d41396/proto/coreum/asset/v1/asset.proto#L40)):*
> ```protobuf
>   string id = 1;
>   string creator = 2;
> }
> ```

it complains bout empty line. and also the equal option seems unnecessary

**Review by wojtek-coreum — APPROVED** [review]
> Reviewed 2 of 4 files at r3, 2 of 2 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/coreumfoundation/coreum/274)* status: all files reviewed, 3 unresolved discussions (waiting on @dzmitryhil, @miladz68, @wojtek-coreum, and @ysv)

**Comment by dzmitryhil** [comment]
The PR is close the new will be with the protos and implementation

**Pull request closed without merging**